### ISSUE TITLE
Custom data + negative samples

### ISSUE THREAD

**Issue by ercz16** [facebookresearch/SEAL_OGB#46]
Hello professor Zhang,
I'm experimenting with a custom dataset using the InMemoryDataset class. My link prediction task involves labeling true edges as those going from chemicals to diseases with a negative correlation (meaning chemical treats disease). My sample graph has some 30k nodes and around 1M edges. If the edge is a negative correlation from chem to disease then it is labeled true. All other edges are labeled false. So the dataset contains a list of edges and a list of corresponding labels, plus node features list and edge features list.

My question regards how the negative sampling works in your code. It appears that negative sampling is done to produce instances for the model recognize as false. But I don't know if my dataset makes sense for this code because it already has false labels. It looks like the code here is expecting all the input edges to be true and then generating negative samples to supplement them. However, my data contains both true and false edges from my data preparation. Do I need to restructure my input dataset? For example, only include true/positive edges in the edge input list ([x, y], [a, b] ... ) and let split_edge dict separate pos/neg edges.

Or is it possible to manually let the model recognize the negative edges that I already labeled? Maybe just sort the neg/pos edges into the split_edge dict manually. I'm also trying to keep a pos/neg ratio of 1:10.

Last question is does this implementation keep node embeddings limited to the training data and not used again in test data to possibly prevent data leakage? My project leader wants to confirm this part.

Any help is appreciated!

**Comment by ercz16** [comment]
Update:
I realized the code supports existing negative edges if already given in split_dict.
Also if I want to incorporate my edge features (list of vectors) into training do I add it in this line? https://github.com/facebookresearch/SEAL_OGB/blob/ea01ef509df1a90bc246e4e9828e192eeec4289c/seal_link_pred.py#L170
I didn't see anywhere else where edge_attr (features) is used.

**Comment by Xi-yuanWang** [comment]
Hello ercz16, 

"the code supports existing negative edges if already given in split_dict." Yes, then no negative edge sampling is needed.

"does this implementation keep node embeddings limited to the training data and not used again in test data to possibly prevent data leakage?" Could you please provide more details on it? Currently, our model produces new node embeddings in each forward process as long as you do not use `--train_node_embedding` option.

"if I want to incorporate my edge features (list of vectors) into training do I add it in this line", Yes. You can change the model here to some models that can take edge feature as input. Like GINE https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.GINEConv.html#torch_geometric.nn.conv.GINEConv
